apache/cloudstack
# Multiple networks support for vms in advanced zone with security group (and kvm support)

## Description



In advanced zone with security groups, there can be multiple shared networks with security groups enabled. However, it is only possible to create vm with only one network. When we try to create vm with multiple networks, it gives error "Only support one network per VM if security group enabled".

Operations succeed if we create a vm with one network and add other networks to the vm. However, the new nics do not work as there are no network rules for them applied on hypervisor.

This PR will enable the functionality on KVM hypervisors, including the support on
(1) API to support vm with multiple networks for KVM
(2) UI to create vm with multiple networks for KVM
(3) network rules applied on all nics (when create a vm, and add/remove nics)
(4) network rules for all secondary ips on all nics

This PR includes the other two bug fixes: #3635 and #3636 
This PR also included a script file for integration test.

Fixes: #3045 #3568 







## Types of changes


## Screenshots (if appropriate):

## How Has This Been Tested?




environment:
ubuntu 18.04 / python2
advanced zone with security groups

check network rules for vm (eg i-10-20-VM)
- ipset list i-10-20-VM
- ipset list i-10-20-VM-6
- iptables-save | grep i-10-20
- ebtables-save | grep i-10-20

vm actions:
- create a vm with multiple networks
- add new nic to vm
- remove nic from vm
- add secondary ips to vm
- remove secondary ips to vm
- reboot vm
- migrate vm
- stop vm
- start vm
